packages/babel-parser/src/parser/statement.ts

```diff
@@ -314,6 +314,42 @@ export default abstract class StatementParser extends ExpressionParser {
     return false;
   }
 
+  startsUsingForOf(): boolean {
+    const lookahead = this.lookahead();
+    if (lookahead.type === tt._of && lookahead.end - lookahead.start === 2) {
+      // using of ...
+      const originalState = this.state;
+      // @ts-expect-error: Second lookahead
+      this.state = lookahead;
+      const lookaheadAhead = this.lookahead();
+      this.state = originalState;
+      if (
+        lookaheadAhead.type === tt._of &&
+        lookaheadAhead.end - lookaheadAhead.start === 2
+      ) {
+        // using of of ...
```

Comment:
In the worst scenario, we have to do triple lookahead to differentiate between `using of of of` and `using of of`.

@rbuckton Maybe we can forbid `of` as a using binding in `for-of`? So `for (using of` must be a for-LHS and we don't have to peek more.

Reply:
Isn't it the same as `let of of of`? Or is `of` disallowed in that case?

Reply:
> for ( [lookahead ∉ { let, async of }] [LeftHandSideExpression](https://tc39.es/ecma262/#prod-LeftHandSideExpression)[?Yield, ?Await] of [AssignmentExpression](https://tc39.es/ecma262/#prod-AssignmentExpression)[+In, ?Yield, ?Await] ) [Statement](https://tc39.es/ecma262/#prod-Statement)[?Yield, ?Await, ?Return]

for-of has disabled `let` and `async` as LHS so `for(let of of` must start a for-of with a let-binding `of`. But `using` comes after for-of so we have to recognize `for(using of of` as for-of LHS.

Reply:
Per https://github.com/tc39/proposal-explicit-resource-management/issues/107, I will disallow `of` as a for-using-of binding name, hopefully we will get rid of extra lookahead, too.

Reply:
There is now a [PR](https://github.com/tc39/proposal-explicit-resource-management/pull/108) that addresses this in the proposal repository.